### PR TITLE
Develop quad/hex mesh support [FEniCS]

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Checklist for prospect students for Google Summer of Code 2017:

- [x] I contacted NumFOCUS mentors before writing my proposal
- [x] I showed you my contribution (it can be any form: proof-of-concept project idea, some sample code or just a link to your commits from other project)
- [x] I linked to my sample contribution from the proposal
- [x] I linked to my opened issues in numfocus/gsoc repository from the proposal

It would be great if FEniCS mentors could help me to improve this proposal. Thank you!
my opened issue #179

### CODE REVIEW

**Comment by kain88-de** [comment]
@martinal

File: 2017/proposals/yashchuk-ivan.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+* Existing code evaluation and improvement.
+
+### July 31st - August 4th, **Begin of Phase 3**
+
+* Continue working with DOLFIN.
+
+### August 7th - August 11th
+
+* Finalize linking everything to DOLFIN.
```
Comment (martinal) [inline 2017/proposals/yashchuk-ivan.md]: You have sketched a fairly linear workflow where you first implement the FIAT/FFC interface, then update other parts of code generation in FFC, then link it up with DOLFIN. For the bulk of the work this is the right order of things, _but_ i think involving the entire pipeline including DOLFIN as early as possible will be more productive.

File: 2017/proposals/yashchuk-ivan.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+### May 4th - May 30th, **Community Bonding Period**
+
+* Familiarize myself completely with FEniCS functionality and architecture.
+* Be in constant touch with my mentor and the FEniCS community.
+* Thus with the help of my mentor I will become absolutely clear about my future goals, the final implementations that need to be done as well as the approach that I will follow.
+
+### May 30th - June 3rd
+
+* Start working on building an interface between FFC and FIAT for quad/hex mesh.
```
Comment (martinal) [inline 2017/proposals/yashchuk-ivan.md]: Having this script work would be a good first milestone:
```
from dolfin import *
mesh = UnitQuadMesh(mpi_comm_world(), 1, 1)
print(assemble(1.0*dx(mesh)))
```
Running this today will lead you straight to the place in FFC where the FIAT/FFC interface needs to be implemented for Q elements. Using `UnitHexMesh` instead will allow testing hexahedral cells as well.

File: 2017/proposals/yashchuk-ivan.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+### June 5th - June 9th
+
+* Testing and improving the interface.
```
Comment (martinal) [inline 2017/proposals/yashchuk-ivan.md]: Another major milestone will be to define function spaces.

This is also part of the work to interface FIAT/FFC but introduces the need for generation of code for dofmaps. To run the code
```
from dolfin import *
mesh = UnitQuadMesh(mpi_comm_world(), 1, 1)
V = FunctionSpace(mesh, "P", 1)
f = Function(V)
```
dofmap generation needs to be fully functional. Dofmaps are a crucial part of the puzzle.

Early June you should be aware of what's needed to get the above running.
To have it fully working at this point may be optimistic though,
it depends a lot on how reusable some of the existing code turns out to be.

File: 2017/proposals/yashchuk-ivan.md (hunk 1 above)
Comment (martinal) [inline 2017/proposals/yashchuk-ivan.md]: A suggestion for high-level milestones, each repeated for meshes of quadrilateral and hexahedral cells:

* Assemble the functionals 1*dx and x[0]*dx
* Define a FunctionSpace
* Interpolate an Expression into a Function
* Assemble the functional f*dx
* Assemble a linear form f*v*dx
* Assemble a bilinear form u*v*dx
* Project an Expression into a Function
* Solve a Poisson problem

The first three points here hide the most complexity and will take the most time to complete.

File: 2017/proposals/yashchuk-ivan.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+### August 28th - August 29th, **Submit final work**
+
+* Have the PR merged.
+* Discussion and evaluation of the work done.
+
+## Future works
+
+I plan to do my PhD on a topic related to FEM, and I think FEniCS will be one of the tools I will use.
+I will try to keep contributing to FEniCS with issues and PRs, and remain an active part of the community.
```
Comment (martinal) [inline 2017/proposals/yashchuk-ivan.md]: This is an excellent plan, and this project will make you understand FEniCS on a deeper level than most.

**Comment by IvanYashchuk** [comment]
Thank you very much for the review, Martin! I will revise the weekly topics and do the changes accordingly

**Comment by IvanYashchuk** [comment]
I have updated the proposal

**Pull request merged**